Tasks chaining side panel

joelclimbsthings opened the pull request:
-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adding another side-bar panel triggered by clicking the "Finish setup" button on the Activity panel. This will display on all non-Homescreen pages where the activity panel appears. 

Closes #32152  .

### How to test the changes in this Pull Request:

1. Checkout branch on fresh site. 
2. Perform the follow steps for **each iteration of the task list (control / experiment 1 / experiment 2)**. In order to toggle those experiments, use the WCAdmin Test Helper Plugin after opting into Marketing.
3.  Go through a few of the tasks on the Homescreen to ensure they function as expected, including the first one that triggers the OBW. You don't necessarily need to complete them, so that you can continue to test. Ensure that those that don't navigate to a non-reactified page don't have a page refresh.
4. Go to another reactified page that is not the Homescreen, that shows the Activity Panel, such as "_Customers_." 
5. Click the "Finish Setup" button, and click on the same tasks to ensure that they function as expected. For most tasks this still should not trigger a page refresh, same as Homescreen.
6. Go to an embed page such as "_Orders_." Click the "Finish Setup" button and ensure that the task list appears as it should. 
7. Trigger the tasks from here as well, and ensure that they work. This time most tasks will direct to the correct task on the Homescreen, and will have a page refresh. 
8. Switch to any untested experiment (if you haven't finished) and return to step #3.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?


### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Basic tasks panel loaded from finish setup button

Commit:
Adding getFile util to navigation package

Commit:
Adding support to default tasks click behavior for non-homescreen pages

Commit:
Tweaking task click logic to allow no page refresh on WCA pages for sidebar task list

Commit:
Adding basic styles for experiment UI

Commit:
Removing orphaned code

Commit:
Adding changelog

Commit:
Converting small setup tasks panel component to typescript

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33124/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Merge branch 'trunk' into add/32152-tasks-side-panel` 1cd273b26ee03453da77ec1d847374ed5a2be44b
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

joshuatf commented on plugins/woocommerce-admin/client/tasks/task-list-item.tsx:
```diff
@@ -133,16 +135,30 @@ export const TaskListItem: React.FC< TaskListItemProps > = ( {
 				window.location.href = actionUrl;
-			} else {
-				( getHistory() as History ).push(
+				return;
```
The logic between here and the two column task list item looks roughly identical.  Do we have plans to remove one of these at some point or could we dry these up a bit?

joelclimbsthings replied:
It was everything I could do to _not_ give into the temptation to factor that logic out. The primary reason I didn't was that this is an experiment that is already in progress (or almost), and we'll presumably be removing all but one of the iterations before any of these changes actually go live? 

I'm not 100% confident that my assumptions there are correct, though.

joshuatf replied:
That makes sense.  If we go forward with the router/navigation utility, that could reduce some of this, right?

joelclimbsthings replied:
Yup, good call!

joshuatf commented on plugins/woocommerce-admin/client/tasks/task-list-item.tsx:
```diff
@@ -133,16 +135,30 @@ export const TaskListItem: React.FC< TaskListItemProps > = ( {
+				return;
+			}
+
+			if ( ! isWCAdmin( window.location.href ) ) {
```
Thinking about this beyond tasks, this would probably be useful as a general utility for all links.  Some method to handle navigation for different link types.

joelclimbsthings replied:
It did occur to me that it may make sense to have a utility that navigates to a page, utilizing the router if appropriate, or redirecting if not (as we're doing here). Is that more-or-less what you had in mind @joshuatf ?

joshuatf replied:
Exactly!  This seems to be a pretty common occurrence and it would be great to have both a utility and test around it for use in all our components.  The navigation package seems like a good candidate for its home.

joelclimbsthings replied:
Consider it done! Elaborated in my [other comment](https://github.com/woocommerce/woocommerce/pull/33124#issuecomment-1137877412), so let me know what you think.

joshuatf commented on plugins/woocommerce-admin/client/two-column-tasks/sectioned-task-list.scss:
```diff
@@ -127,3 +127,25 @@
+		font-weight: 400;
+	}
+
+	.woocommerce-task-progress-header {
```
This looks good.  The completed state in the 2 experiments feels a bit odd to me; not sure if this is intentional or not.  cc @jarekmorawski 
<img width="449" alt="Screen Shot 2022-05-24 at 10 00 08 AM" src="https://user-images.githubusercontent.com/10561050/170055013-b18c2fc3-e7fa-4e58-bac0-058c4d106ed0.png">
<img width="457" alt="Screen Shot 2022-05-24 at 9 58 09 AM" src="https://user-images.githubusercontent.com/10561050/170055016-de655f7a-90fa-469d-96db-e2fb892193e1.png">

jarekmorawski replied:
Hiya! What specifically looks odd to you? When all tasks are completed, we should show the congratulatory card with the CES emoji button. I think it was implemented as part of the checklist experiment in the home screen.

joshuatf replied:
The "Welcome to Site Name" piece looks odd to me here.  The success screen looks good though.

jarekmorawski replied:
Oh, I totally missed the header 🙈 In the design, the header has a different appearance in the side panel; the fonts and paddings are smaller to bring out the progress bar and the tasks. When completed, the header is hidden completely so that only the celebratory card is visible.

<img width="1504" alt="image" src="https://user-images.githubusercontent.com/79307566/170072245-75eb4b88-9869-41a9-a727-aca8fcf36ba5.png">

When someone brings back the list after completing it, we should hide the header altogether.

jarekmorawski replied:
Note that in the designs the section titles have smaller fonts (14 px vs 20 px). It concerns both the task list and `Things to do next.` It's like a mobile-friendly version of the home screen.

joshuatf replied:
Nice, thanks, @jarekmorawski!

joelclimbsthings replied:
After taking a closer look at the design, I've applied the altered styles to task list when displayed in the sidebar in https://github.com/woocommerce/woocommerce/pull/33124/commits/e1834e3ce116cd4c8135cb81987199de8d1a6d88. 

I also ensured that the progress header is not displayed when the list is complete.

Let me know if I missed anything from this thread!

joshuatf reviewed (commented):
Looking great and testing well!  Left a potential option to dry things up and one question around design.

Commit:
Revising sidebar tasks styles based on design

Commit:
Making displayProgressHeader conditional on task list not being complete

Commit:
Adding small margin next to checkmark badge on task list

joelclimbsthings commented:
Thanks for the review @joshuatf ! I addressed the UI / styling issues, and responded to your other comments. Let me know if you have any additional thoughts on those.

Commit:
Creating utility to navigate to url using best method, refactoring task click handlers to utilize

Commit:
Moving isWCAdmin utility to navigation package

Commit:
Fixing JSdoc optional params

joelclimbsthings commented:
Hey @joshuatf , I've added the `navigateTo` utility to the `@woocommerce/navigation` package. I had to touch a handful of files since this also required relocation the `isWCAdmin()` utility to this package, and had to remedy the few files that were using it. 

Since the diff for this PR is getting a little heavy, it might be best to scope these two commits out in isolation: 
https://github.com/woocommerce/woocommerce/pull/33124/commits/831ca012c8c70645ef381cc21518fccb33aac2bf
https://github.com/woocommerce/woocommerce/pull/33124/commits/4fb6429e8913b339f72e45a4cf776dfd23f9f542

Commit:
Adding tests for navigateTo utility

joshuatf commented on packages/js/navigation/src/index.js:
```diff
@@ -266,6 +267,42 @@ export const addHistoryListener = ( listener ) => {
+ * @param {*} url - the url to test
+ * @return {boolean} true if the url is a wc-admin URL
+ */
+export const isWCAdmin = ( url = window.location.href ) => {
```
Nice!  I like the default of `window.location.href`.

joshuatf commented on packages/js/navigation/src/index.js:
```diff
@@ -266,6 +267,42 @@ export const addHistoryListener = ( listener ) => {
+
+	const newPath = path ? path : getNewPath( {}, url, {} );
+
+	if ( ! isWCAdmin() ) {
```
Do we need to pass `newPath` here?  Otherwise it seems like it will check the current page instead of the URL in question.

joelclimbsthings replied:
Good thought @joshuatf . For my current use case it was actually necessary to test the current URL, since I'm handling the case of navigating _from_ an embedded page to a WCAdmin page. That said, I should be checking both the origin and destination, which I've added.

joshuatf commented on packages/js/navigation/src/index.js:
```diff
@@ -266,6 +267,42 @@ export const addHistoryListener = ( listener ) => {
+ * @param {string} [args.url]  - Relative or absolute url to navigate to.
+ * @param {string} [args.path] - Path to navigate to.
+ */
+export const navigateTo = ( { url, path } ) => {
```
I find the `path` and `url` params a bit confusing (especially since URLs can be relative).

Is this in here because we might have a fully qualified URL that is actually a WCA page?  Is there any opportunity to combine these into a single argument or do you think that complicates the functionality too much?

joelclimbsthings replied:
@joshuatf This is definitely true, and I've ended up revamping the utility to only accept the url parameter. That said, this may require a bit more thought since the way(s) that we're handling these links is a bit all over the place. 

For example, the `actionUrl` for the tasks can be either an absolute URL, or actually just just `path` parameter. The second of which just assumes that it's a WCAdmin page, so we don't have the ability to get a non-wcadmin page via that parameter.

Right now in the `navigateTo` utility I've settled for assuming either an absolute URL or a wc-admin path (starting with `admin.php` as is done with the `getNewPath()` util), with the assumption that it's relative to `/wp-admin/`. I also left that last bit configurable, but I'm still a bit uncertain how we want to approach all of this. Further work should likely move to another PR, but it's been nice to utilize it for the tasks!

joshuatf commented on packages/js/navigation/src/test/index.js:
```diff
@@ -260,3 +262,114 @@ describe( 'getSetOfIdsFromQuery', () => {
 	} );
 } );
+
+describe( 'isWCAdmin', () => {
```
Love all these tests ❤️

joshuatf commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php:
```diff
@@ -407,7 +407,7 @@ public function get_json() {
 			'tasks'                 => $tasks_json,
 			'eventPrefix'           => $this->prefix_event( '' ),
-			'displayProgressHeader' => $this->display_progress_header,
+			'displayProgressHeader' => $this->display_progress_header && ! $this->is_complete(),
```
I don't think we want to hide this everywhere when tasks are complete, probably just in the sidebar.

The `Welcome to {siteTitle}` feels a little weird inside the task list anyway.  I wonder if we could just move that portion outside the task list on the homepage.  cc @louwie17 for your thoughts.

If that's not an option, we could add an `is-complete`  class to the task list so we can hide it via CSS in the sidebar.

joelclimbsthings replied:
I assumed this would apply to both the side bar as well as the Homescreen, since I wasn't sure it made sense to display the progress in general after everything was complete. But perhaps not? 

I'm not sure of all the requirements around the progress bar, but glad to make this change. At this point further changes will have to wait until Monday, though.

louwie17 replied:
I agree with @joshuatf here, I think the `displayProgressHeader` here should not be dependant on is_complete and that part being handled on the front-end instead. In conjunction with moving the `Welcome to {siteTitle}` to the home screen.

The welcome site title header was part of the progress header when the completed header wasn't insight yet. Now that we do have the completed header, we probably only want to display the `Welcome to site` header when a task list is hidden, which can be handled on the homepage.
We might want to double check that logic with @jarekmorawski

joelclimbsthings replied:
I'm a little confused on the requirements here, partially since we are juggling so many different experiments/contexts here at this point. 

In terms of moving the title to the Homescreen, it's my understanding that it _is_ meant to be displayed on the sidebar according to the designs. Keep in mind that it has differing states, depending on tasks completed, which the figma clearly displays within the sidebar: 

![image](https://user-images.githubusercontent.com/444632/171045388-be3889c9-2d12-4910-9c28-b255ce3937eb.png)

At this point I think the only condition which we're concerned with is showing this title (and progress bar?) within the sidebar after the tasks are complete. I'm wondering whether something along the lines of @joshuatf 's suggestion on simply adding a class `is-complete` and using styling here would be "good enough." We could start passing around more props as well, but so far it hasn't been necessary to discern the sidebar state in those components and I'm a bit reluctant.

joshuatf reviewed (commented):
Really great work in here.  Love the navigation utility and tests around it 💯 

Left a few follow-up comments/questions.

Commit:
Revamping navigateTo utility to be more general and only require one parameter

Commit:
Fixing tests and ensuring absolute actionUrls are handled correctly

Commit:
Removing is_complete from displayProgressHeader boolean on task lists

joelclimbsthings commented:
@joshuatf Given the [conversation we've had in various channels](p1653583936663609-slack-C01CMSA1VHS), I'd like to aim to get this merged shortly and we can bring up the progress title as a point of conversation during the call tomorrow. I've reverted the `displayProgressHeader` change on the back-end, and rebased to resolve conflicts.

joshuatf commented on packages/js/navigation/src/index.js:
```diff
@@ -266,6 +266,39 @@ export const addHistoryListener = ( listener ) => {
+ * @param {string} args.url                    - Relative path or absolute url to navigate to
+ * @param {string} [args.directory=/wp-admin/] - Path directory, defaultsl to admin path.
+ */
+export const navigateTo = ( { url, directory = '/wp-admin/' } ) => {
```
We can't depend on `directory` being `wp-admin` since sites can customize this.

I believe `getAdminLink` is what we're currently using.

joelclimbsthings replied:
Good call! I should have remember, as a very long time ago I did this very thing on my own WordPress installation. I've removed the logic around the directory.

joshuatf commented on plugins/woocommerce-admin/client/tasks/task-list-item.tsx:
```diff
@@ -130,16 +126,15 @@ export const TaskListItem: React.FC< TaskListItemProps > = ( {
-			} else {
-				getHistory().push( getNewPath( {}, actionUrl, {} ) );
-			}
+			navigateTo( {
```
Can this not be changed to `navigateTo( actionUrl )`?  Do we need to have the `getNewPath( {}, actionUrl, {} ),` piece in here or can that logic be contained within the navigation function?

joelclimbsthings replied:
The issue here is that we currently have `actionUrl`'s of simply `/setup-wizard`, as well as full absolute URLs (as a side note we might want to revisit the format of the `actionUrl` for this reason, or at least document it somewhere clearly). Since this is a bit of a niche/non-standard format, it seemed cleaner to keep this logic here. I'm currently thinking that the cleanest implementation of `navigateTo` only accepts an absolute URL, or one relative to the admin directory, which works well with our existing utilities like `getNewPath()`. 

Thinking about this a bit more, we _could_ just assume that anything that isn't absolute and does not start with a php file is a wcadmin path, perhaps? I'm reluctant to start adding a bunch of regex and various conditions to try to parse several different potential formats, but that may be a good middle-ground. That's what I've now done in https://github.com/woocommerce/woocommerce/pull/33124/commits/dfd9f77b55fcb3c7144ede9e8574a5f86970df5c. Let me know what you think @joshuatf

joshuatf replied:
Ya, that makes sense.  We really should not have introduced the pattern of using partial WCA URLs as this is coming back to bite us now.

But I like the short term regex check.  This feels like a good intermediary until we can resolve the bad partial URLs.

joshuatf reviewed (commented):
Thanks for the changes, Joel.  I'm good holding off on moving the "Welcome to ..." portion to another PR as long as we make sure to follow up prior to code freeze.

Added a couple comments around the navigation function I think we need to address prior to merging.

Commit:
Refactoring progress title outside of tasks header, minor style changes

Commit:
Displaying finish setup button during task completion

Commit:
Resolving issue with progress title displayed on task screens

Commit:
Removing brittle wcadmin directory from navigateTo util

Commit:
Refactoring navigateUrl util to support various types of url formats

joelclimbsthings commented:
Hey @joshuatf , appreciate your efforts here! I've responded to your inquiries above, refactoring a few things with the progress title and the `navigateTo()` utility. See my above comments for more details there. 

As mentioned earlier our current approach to links across the application, including the `actionUrl` we're using for tasks, could likely use a little more thought and potential discussion. I'm curious what you're thoughts are thus far.

I'm still planning to add a property to the tracks events as was mentioned earlier today, but I may prefer just to throw that in another PR since this is already touching 27 files and is going to start accruing more conflicts. Either way I can look at that tomorrow.

joshuatf reviewed (approved):
Thanks for all the persistence here, @joelclimbsthings!  Things are looking great and it's really nice to have this super valuable util now.

It would be great if we could also create a follow-up to tackle adding this to the inbox notes as this would be really handy there as well.
